### PR TITLE
Fix bug: Advancing wrong pointer 

### PR DESCRIPTION
**Pull request by michael-weigelt**

### CODE REVIEW

**Review by basvandijk — APPROVED** [review]

**Comment by dsd-at-dfinity** [comment]
short high-level descr in the PR would be nice also in these cases. :-)

File: lib/Wasmtime.hs [hunk 1]
```diff
@@ -997,13 +998,13 @@ newtype MemoryType = MemoryType {unMemoryType :: ForeignPtr C'wasm_memorytype_t}
...
+newMemoryType ::
   -- | Minimum number of memory pages.
-  Word64 -> 
+  Word64 ->
   -- | Optional maximum of memory pages.
-  Maybe Word64 -> 
-   -- | 'True' means memory is 64 bit flag and 'False' means 32 bit.
-  Bool -> 
+  Maybe Word64 ->
+  -- | 'True' means memory is 64 bit flag and 'False' means 32 bit.
```
Comment (dsd-at-dfinity) [inline lib/Wasmtime.hs:1006]: btw, would it make sense to have an ADT for this instead of true/false?

File: lib/Wasmtime.hs [hunk 2]
```diff
@@ -997,13 +998,13 @@ newtype MemoryType = MemoryType {unMemoryType :: ForeignPtr C'wasm_memorytype_t}
 
 -- | Creates a descriptor for a WebAssembly 'Memory' with the specified minimum number of memory pages,
 -- an optional maximum of memory pages, and a 64 bit flag, where false defaults to 32 bit memory.
-newMemoryType :: 
+newMemoryType ::
   -- | Minimum number of memory pages.
-  Word64 -> 
+  Word64 ->
```
Comment (dsd-at-dfinity) [inline lib/Wasmtime.hs:1003]: stupid question: would a newtype do the job here? So we would have `BasePointer` and `Offset` ....

File: lib/Wasmtime.hs [hunk 3]
```diff
@@ -1122,7 +1123,7 @@ readMemoryAt ctx mem offset len = do
       else do
         res <- unsafeWithMemory ctx mem $ \mem_data_ptr mem_size ->
           BI.create mem_size $ \dst_ptr ->
-            BI.memcpy (advancePtr dst_ptr (fromIntegral offset)) mem_data_ptr mem_size
+            BI.memcpy dst_ptr (advancePtr mem_data_ptr (fromIntegral offset)) mem_size
```
Comment (dsd-at-dfinity) [inline lib/Wasmtime.hs:1126]: I meant to place the above comment here. 

Would it make sense to have a separate (newtype?) for offset and length respectively?

File: lib/Wasmtime.hs (hunk 1 above)
Comment (michael-weigelt) [reply lib/Wasmtime.hs:1006]: I think so. In general, we need to do pass through this C-adapted API and use more haskell idioms in the exposed functions. Thanks for pointing this out!

**Comment by michael-weigelt** [comment]
I have changed some types. TBH, we don't gain much from the bool alternative, because we don't pattern match on them. It is only an improvement for the reader of the type. 

Similarly, while Offsets and Lengths should absolutely be non-negative, the C functions curiously accept Ints, so I have to convert them inline, which is ugly. I think there it's worth it, because no user should be able to read backwards from an offset pointer, or use negative lengths. 

WDYT?

File: lib/Wasmtime.hs [hunk 4]
```diff
@@ -1029,8 +1043,11 @@ getMax mt = unsafePerformIO $
         else Just <$> peek max_ptr
 
 -- | Returns false if the memory is 32 bit and true if it is 64 bit.
-is64Memory :: MemoryType -> Bool
-is64Memory mt = unsafePerformIO $ withMemoryType mt c'wasmtime_memorytype_is64
+is64Memory' :: MemoryType -> Bool
+is64Memory' mt = unsafePerformIO $ withMemoryType mt c'wasmtime_memorytype_is64
+
+is64Memory :: MemoryType -> WordLength
+is64Memory mt = if is64Memory' mt then Bit64 else Bit32
```
Comment (basvandijk) [inline lib/Wasmtime.hs:1050]: What about calling this one:
```suggestion
wordLength :: MemoryType -> WordLength
wordLength mt = if is64Memory' mt then Bit64 else Bit32
```
? And then removing the ' from `is64Memory'`.

File: lib/Wasmtime.hs [hunk 5]
```diff
@@ -179,6 +181,17 @@ import Foreign.Storable (Storable, peek, peekElemOff, poke, pokeElemOff)
 import System.IO.Unsafe (unsafePerformIO)
 import Type.Reflection (TypeRep, eqTypeRep, typeRep, (:~~:) (HRefl))
 
+--------------------------------------------------------------------------------
+-- Typedefs
+--------------------------------------------------------------------------------
+
+type Offset = Word64
+
+type Length = Word64
```
Comment (basvandijk) [inline lib/Wasmtime.hs:190]: Super nitpicking but wasmtime calls the length of the memory "size" so what about:
```suggestion
type Size = Word64
```
?

**Review by basvandijk — APPROVED** [review]

File: lib/Wasmtime.hs (hunk 4 above)
Comment (michael-weigelt) [reply lib/Wasmtime.hs:1050]: yes, but we are straying from the nomenclature of wasmtime..

**Pull request merged**